pkg/controller/common/reconcile.go

```diff
@@ -0,0 +1,183 @@
+package common
+
+import (
+	"fmt"
+	"reflect"
+
+	ign3types "github.com/coreos/ignition/v2/config/v3_4/types"
+	mcfgv1 "github.com/openshift/api/machineconfiguration/v1"
+	"github.com/openshift/machine-config-operator/pkg/daemon/constants"
+	"k8s.io/klog/v2"
+)
+
+// IsRenderedConfigReconcilable checks the new rendered config against the old
+// rendered config to make sure that the only changes requested are ones we
+// know how to do in-place. If we can reconcile, (nil) is returned. Otherwise,
+// if we can't do it in place, the returned string value includes the
+// rationale.
+//
+// We can only update machine configs that have changes to the files,
+// directories, links, and systemd units sections of the included ignition
+// config currently.
+func IsRenderedConfigReconcilable(oldConfig, newConfig *mcfgv1.MachineConfig) error {
+	return IsComponentConfigsReconcilable(oldConfig, []*mcfgv1.MachineConfig{newConfig})
+}
+
+// IsComponentConfigsReconcilable checks each individual component
+// MachineConfig against the old rendered config. This is slightly more
+// efficient because we only have to process the old config once. If an error
+// occurs, the name of the component config will be included in the error
+// message.
+func IsComponentConfigsReconcilable(oldConfig *mcfgv1.MachineConfig, newConfigs []*mcfgv1.MachineConfig) error {
+	// The parser will try to translate versions less than maxVersion to maxVersion, or output an err.
+	// The ignition output in case of success will always have maxVersion
+	oldIgn, err := ParseAndConvertConfig(oldConfig.Spec.Config.Raw)
+	if err != nil {
+		return fmt.Errorf("parsing old Ignition config from machineconfig %q failed: %w", oldConfig.Name, err)
+	}
+
+	// Go through each component config and determine if it is reconcilable.
```

Comment:
question: is there any possible scenario where the merged ignition is reconcileable but some individual one isn't? For example:

I have machineconfig 1 that configures a disk partition during install time
I have machineconfig 2 that sets up FIPS

When merged the are reconcileable, but if I check them individually, 1 sub-machineconfig technically "changed FIPS" and 1 sub-machineconfig technically "changed disk partition" such that they both fail this check?

Reply:
This is a great question!

I was thinking about that and even ran into that scenario during my e2e testing. The solution was to not care about whether a rendered MachineConfig can be reconciled until after the MachineConfigPool actually has a rendered MachineConfig applied to it. In other words, until `Spec.Configuration.Name` is set on the MachineConfigPool, we don't check for reconciliation because there is nothing to check against; so we just produce a rendered MachineConfig. Once the MachineConfigPool has a rendered MachineConfig, then we can check both the component and final rendered configs against the the pools' current status.

There is an e2e job that runs with FIPS enabled that I will run against this PR in order to check for that specific edge-case.

Reply:
Ahhh, that's a point I hadn't considered. I'll do a bit more testing for that particular scenario so I'll know if it affects us or not. My gut feeling is that it shouldn't affect us, but I could be wrong.

Reply:
Aha!! The FIPS job I described failed precisely for the reason you just mentioned!

Reply:
> Using your example, we'd end up with a rendered MachineConfig that sets FIPS to enabled, but we'd still have the old one that has it marked disabled.

FIPs needs to be set at install time and install time only, so I think in theory, no currentConfig should ever have FIPs:disabled if you enable a FIPs cluster. If it changed it would be a bug in the installer(?) or bootstrap MCO

Put another way, this works today in the daemon, so if we do the exact same check in the controller, it should pass. Unless you are thinking of another scenario I am missing :thinking:

Reply:
Basically, any non-reconcileable machineconfig changes must be install time, such that they are always present in any "currentConfig" field we check.

Reply:
I think I misinterpreted the scenario you described based upon what I was seeing from the FIPS test. I made a slight change that will only examine the component configs if reconciliation fails against a generated config. So far, the FIPS test is showing that my change may have been successful :smile:.